🆕 (android) Add a page detailing the camera orientations

lnishan opened the pull request:
Adds a page detailing the orientations involved when using the camera2 API to display camera preview
and capture jpeg output images.

**Please describe what your Pull Request does**

The PR adds a page and illustrations that detail orientations that are involved when creating a Android camera application

**Please tag each issue this Pull Request resolves**

- Resolves #

Commit:
🐛 (android) Updating Window Manager Page for Android

Commit:
🐛 (android) Updating Window Management Page

Commit:
🐛 (android) Updating window management page

leopic committed:
Ensure GitHub link points to right organization

leopic committed:
Ensure GitHub link points to right organization

Commit:
🐛 (android) Updating Window Managemet Page

dependabot[bot] committed:
Bump prismjs from 1.20.0 to 1.21.0

Bumps [prismjs](https://github.com/PrismJS/prism) from 1.20.0 to 1.21.0.
- [Release notes](https://github.com/PrismJS/prism/releases)
- [Changelog](https://github.com/PrismJS/prism/blob/master/CHANGELOG.md)
- [Commits](https://github.com/PrismJS/prism/compare/v1.20.0...v1.21.0)

Signed-off-by: dependabot[bot] <support@github.com>

lnishan committed:
🆕 (android) Add a page detailing the camera orientations

Adds a page detailing the orientations involved when using the camera2 API to display camera preview
and capture jpeg output images.

lnishan commented:
```
> chromeos.dev@1.0.0 prettier:lint /Users/lnishan/chromeos.dev
> prettier -c './**/*.{html,md,js,json}'

Checking formatting...
lib/collections/news-stories.js[14:30:50] Using gulpfile ~/chromeos.dev/gulpfile.js
[14:30:50] Starting 'sass'...
package-lock.json[14:30:54] Finished 'sass' after 3.44 s
pages/_generated/components.htmlpages/en/android/camera-orientation.md
    7:3-7:23  warning  'tags[0]': should be equal to one of the allowed values  content-models
    8:3-8:19  warning  'tags[1]': should be equal to one of the allowed values  content-models
    9:3-9:22  warning  'tags[2]': should be equal to one of the allowed values  content-models
  10:3-10:21  warning  'tags[3]': should be equal to one of the allowed values  content-models
  11:3-11:24  warning  'tags[4]': should be equal to one of the allowed values  content-models
  12:3-12:23  warning  'tags[5]': should be equal to one of the allowed values  content-models
  13:3-13:21  warning  'tags[6]': should be equal to one of the allowed values  content-models
  14:3-14:24  warning  'tags[7]': should be equal to one of the allowed values  content-models
```

Note that this commit failed presubmit check, but I'm not sure what I did wrong here.

lnishan commented:
Ok I see. I need to add tags to
https://github.com/chromeos/chromeos.dev/blob/trunk/lib/linting/schemas/definitions/tags.js

Let me do that in a bit.

lnishan committed:
🐛 (code) Fix tags in camera-orientation

Fixes the tags in camera-orientation by adding the descriptions in the linting config. Also makes
the list of tags shorter.

lnishan committed:
🐛 (code) Add missing comma

Adds a missing comma to a linting config file

lnishan committed:
🐛 (code) Fix style of camera-orientation with prettier

Fixes the style of camera-orientation with `npx prettier --write`

lnishan commented:
Ok I think I may finally have this sorted out. I really don't have much experience with these packages (last time I used nodejs was probably 2.5 years ago!)

So adding a page needs
1. Headers with metadata
2. Tags should be defined and described in `lib/linting/schemas/definitions/tags.js`
3. npx prettier --write './**/*.{html,md,js,json}'

emilieroberts commented:
Hey, I don't see a staging preview which would be handy for reviewing, I wonder if this is a tag thing?

nodejs is a pain in the butt.  On my linux machine I have to run "npx git-cz" instead of git commit.  Could you try that and see if the tags magically work?

joycetoh8 committed:
📝 (android-env) Rename "Android developer options" page

Rename "Android developer options" to "More tools for Android developers" and include flutter
install content.

lnishan committed:
🎨 (android) Apply suggestions from the team

Applies the suggestions from the team, which include a wide range of wording changes. Also reduces
the number of tags to 2.

lnishan commented:
@emilieroberts 

Please help take a look.
Thank you so much for the amazing suggestions :)

leopic committed:
Use correct icon for Spanish's Publish section

Commit:
🐛 (android) Merging in WM changes collab with Ken

Commit:
🐛 (android) Fixing small glitches in formatting

Commit:
🐛 (android) Few nits and edits

lnishan commented:
Hi Emilie,
Have you had chance to take a second look? :)
Not in a hurry to get this published, but just checking in.

lnishan committed:
🐛 (android) Correct description of sensor orientation

Corrects the description of sensor orientation, and a minor edit to wording.

Snugug commented on pages/en/android/camera-orientation.md:
```diff
@@ -0,0 +1,301 @@
+
+Handling orientations in Android camera apps is tricky and needs to take the following factors into consideration:
+
+1. Natural orientation: The display orientation when the device is in the “normal” position for the device design - usually portrait orientation for mobile phones and landscape orientation for laptops.
+2. Sensor orientation: The orientation of the sensor physically mounted on the device.
+3. Display rotation: How much the device is physically rotated from the natural orientation.
+4. Viewfinder size: The size of the viewfinder used for displaying the camera preview.
+5. Image size output by the camera.
```
These don't read like an ordered list to me, but rather like a group of things that could be in any order. If that's true, can we make it an unordered list? If not, can it be rewritten so the items are clearly needed in a specific order?

Thanks!
```suggestion
* Natural orientation: The display orientation when the device is in the “normal” position for the device design - usually portrait orientation for mobile phones and landscape orientation for laptops.
* Sensor orientation: The orientation of the sensor physically mounted on the device.
* Display rotation: How much the device is physically rotated from the natural orientation.
&* Viewfinder size: The size of the viewfinder used for displaying the camera preview.
* Image size output by the camera.
```

lnishan replied:
There is an extra ampersand that I cannot take out here. 

> &* Viewfinder size

Snugug replied:
Removed it, thanks for catching!

lnishan replied:
Awesome. Thanks for the suggestion here :)

Snugug commented on pages/en/android/camera-orientation.md:
```diff
@@ -0,0 +1,301 @@
+
+![Image illustration showing a stretched bugdroid as a result of fitting a portrait-shaped preview into a landscape-shaped viewfinder](/images/android/camera-orientation/surface_view_stretched_scene.png)
+
+You generally want **the aspect ratio (i.e., width/height) of the viewfinder to be identical to the aspect ratio of the source**. If you don’t want to clip the image in the viewfinder - cutting off some of the pixels to fix the display, there are 2 cases to consider:
+
+##### 1. `aspectRatioActivity > aspectRatioSource`
```
```suggestion
You generally want **the aspect ratio (i.e., width/height) of the viewfinder to be identical to the aspect ratio of the source**. If you don’t want to clip the image in the viewfinder - cutting off some of the pixels to fix the display, there are 2 cases to consider, when `aspectRatioActivity` is greater than `aspectRatioSource` and when it's less than or equal to `aspectRatioSource`

##### `aspectRatioActivity > aspectRatioSource`
```

Snugug commented on pages/en/android/camera-orientation.md:
````diff
@@ -0,0 +1,301 @@
+viewfinderWidth = activityHeight * aspectRatioSource;
+```
+
+##### 2. `aspectRatioActivity ≤ aspectRatioSource`
````
```suggestion
##### `aspectRatioActivity ≤ aspectRatioSource`
```

Snugug commented on pages/en/android/camera-orientation.md:
```diff
@@ -0,0 +1,301 @@
+
+![Illustration of a phone with a display rotation of 90 degrees, and an object](/images/android/camera-orientation/texture_view_observer_side.png)
+
+**1. Set the size of the TextureView to be identical to the preview size chosen**
+
+Suppose the preview size you chose is `previewWidth × previewHeight` where `previewWidth > previewHeight` (sensor output is landscape-shaped by nature). When configuring a capture session, one should call [SurfaceTexture#setDefaultBufferSize(int width, height)](<https://developer.android.com/reference/android/graphics/SurfaceTexture#setDefaultBufferSize(int,%20int)>) to specify the preview size ( ...
+
+Before calling setDefaultBufferSize, it’s important that you **also set the size of the TextureView to be `previewWidth × previewHeight`** with [View#setLayoutParams(android.view.ViewGroup.LayoutParams)](<https://developer.android.com/reference/android/view/View#setLayoutParams(android.view.ViewGroup.LayoutParams)>). The reason for this is that TextureView calls `SurfaceTexture#setDefaultBufferSi ...
+
+Now the TextureView may not match the dimensions of the source. In the case of phones, the source is portrait-shaped, yet the TextureView is landscape-shaped owing to the layoutParams you just set. This would result in a stretched previews, as illustrated below,
+
+![Image illustration of a portrait-shaped preview stretched to fit inside a TextureView of the same size of the preview size chosen](/images/android/camera-orientation/texture_view_step_1_identical_size.png)
+
+##### 2. Scale the potentially stretched TextureView back to the original dimensions of the preview
```
Please ensure these items in the process are consistently headers or just bold text in a paragraph, and please don't include numbers if they're headers.

lnishan replied:
Here we're trying to show a 3-step process, so it's procedural.

The problem here is each of the 3 steps needs quite a bit of paragraph, and since it's procedural, I think it makes sense to include numbers to indicate which step we're in.

Let me think how this can be dealt with within the constraints of Markdown.

lnishan replied:
Corrected :) Thanks!

Snugug commented on pages/en/android/camera-orientation.md:
```diff
@@ -0,0 +1,301 @@
+
+- [PreviewView](https://developer.android.com/reference/androidx/camera/view/PreviewView) from [camerax](https://developer.android.com/training/camerax) in [Jetpack](https://developer.android.com/android/jetpack) handles TextureView layout as described above. It configures the transformation with [PreviewCorrector](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-master-dev: ...
+
+##### Note
+
+If you have previously used transformation matrix for TextureView in your code, the preview may not look right on a naturally-landscape device like Chromebooks. Likely your transformation matrix incorrectly assumes the sensor orientation to be 90 or 270 degrees. You may refer to this [commit](https://github.com/android/camera-samples/commit/3d1a254eb018a51ff39ae78d39a9e9e7942a027b) on GitHub for  ...
```
```suggestion
!!! aside.message--note
**Note:** If you have previously used transformation matrix for TextureView in your code, the preview may not look right on a naturally-landscape device like Chromebooks. Likely your transformation matrix incorrectly assumes the sensor orientation to be 90 or 270 degrees. You may refer to this [commit](https://github.com/android/camera-samples/commit/3d1a254eb018a51ff39ae78d39a9e9e7942a027b) on GitHub for a workaround, but we highly recommend that you migrate your app to use the method described here instead.
!!!
```

We have a specific set of markdown fences to render this content as a note that should be used here.

Snugug commented on pages/en/android/camera-orientation.md:
```diff
@@ -0,0 +1,301 @@
+date: 2020-09-23
+weight: 1
+tags:
+  - camera orientation
+  - large screens
```
```suggestion
  - camera
```

Please remove the "large screens" tag as this article doesn't appear to be about utilizing extra space for larger screens. If that's not correct, feel free to keep it. Camera in next comment

Snugug commented on lib/linting/schemas/definitions/tags.js:
```diff
@@ -32,6 +32,7 @@ module.exports = {
     'stylus support': 'For content related to stylus input integration and related stylus interactions',
     'touchscreen support': 'For content related to checking touchscreen availability on devices and handling touchscreen events when appropriate',
     'trackpad support': 'For content related to handling trackpad-specific interactions and gestures',
+    'camera orientation': 'For content related to handling orientations in an Android app with camera functionality',
```
```suggestion
    'camera': 'For content related to handling camera functionality',
```

I think a more generic `camera` tag here would catch more content, and therefore build a more robust collection, than the specific `camera orientation`

lnishan replied:
Thanks for your suggestion!
Is there anyway we can keep "orientation"? I added "large screen" because this is somewhat related (small screens, phones, don't have orientation issues).
This is an important keyword for the article, and I think it makes sense to include it to make the content more discoverable.

lnishan replied:
To be more specific, I'm thinking if we can have 2 tags: "camera" and "orientation". What do you think?

Snugug replied:
I'm open to the idea, but let's. tease it out a little more. There are two different orientations we could be talking about; device orientation and window orientation, which. are somewhat related but independent. Are you recommending one "orientation" tag to cover both, separate tags, or something else?

lnishan replied:
That's a very good question. I think this article is relevant to both.

Hmm... I was going to suggest "orientation" because this can also be added to window-management.
However I see that we already have a "device configuration" tag
> 'device configuration': 'For content related to the device mode (clamshell, tablet, touchscreen), orientation (portrait, orientation), and other configuration changes'
This seems relevant to this article. Can I use "camera" and "device configuration" here?

Snugug replied:
Yah, I think camera and device configuration are good tags for this.

Snugug reviewed (changes requested):
Thanks for all of the work on this new piece of content! It's really appreciated! I have a couple of formatting comments on the markdown. Once resolved, this looks good to me.

lnishan committed:
Replace the 2 cases of SurfaceView with just bullet points

Co-authored-by: Sam Richard <samrichard@google.com>

lnishan committed:
Replace the 2 cases of SurfaceView with just bullet points

Co-authored-by: Sam Richard <samrichard@google.com>

lnishan committed:
Use markdown fence for the note on TextureView

Co-authored-by: Sam Richard <samrichard@google.com>

lnishan committed:
Replace numbered list with bullet point for the orientation overview section

Co-authored-by: Sam Richard <samrichard@google.com>

lnishan committed:
🐛 (android) Change the 3-step process to use bold text

Changes the 3-step process for TextureView to use bold text.

lnishan committed:
🎨 (android) Use '-' for bullet points

Use '-' for bullet points since prettier seems to prefer '-' over '*'.

lnishan commented:
Btw, Sam you might need CLA here, though I'm not certain why.

Snugug commented:
@googlebot I fixed it.

Snugug commented:
> Btw, Sam you might need CLA here, though I'm not certain why.

I don't know what Googlebot is complaining about here? I'm gonna ping the maintainer and see what's going on.

lnishan commented:
@googlebot I fixed it.

lnishan committed:
Use 'camera' as tag for contents related to camera

Co-authored-by: Sam Richard <samrichard@google.com>

lnishan committed:
Use 'camera' tag and remove 'large screens'

Co-authored-by: Sam Richard <samrichard@google.com>

lnishan committed:
🎨 (android) Add 'device configuration' as tag

Add 'device configuration' tag to camera-orientation.md

lnishan committed:
🐛 (code) Remove single quote of the camera tag

Remove single quote of the camera tag per suggestion from prettier

lnishan commented:
I've finished the PRs to modify tags. Thank you so much for working with me on this article, Sam! :) 
(I'm also very impressed by the setup of this project, including the documentations.

Please let me know when you hear back from them. Thanks again!

lnishan commented:
Hmm I'm a little surprised interactive rebase results in this.

Let me just create a new PR.
